### PR TITLE
.travis.yml: publish gh releases only once

### PR DESCRIPTION
**Pull request by samsalisbury**
- Was trying to run 'deploy' step after every vector in the matrix.
- Now runs deploy step after 'make test-unit' only, so we get basic
  checks and quick binary release.

### CODE REVIEW

**Comment by sous-machine-account** [comment]
# LGTM!

No errors, no warnings. :shipit:


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by lachlanh** [comment]
don't know enough about travis config to see how this will change the publishing of the binary. Will mention @nyarly in that case and see if he has an opinion on it.

File: .travis.yml [hunk 1]
```diff
@@ -53,6 +53,7 @@ deploy:
   - artifacts/sous*.deb
   on:
     tags: true
+    condition: '"$TEST_COMMAND" = "make test-unit"'
```
[the remark this answers is not in the document]
Comment (samsalisbury) [reply .travis.yml]: Cool, wasn't aware of that mechanism, looks good, I'll update this. Still would like to generate binaries quicker, but I can appreciate it's safer to wait for all tests to pass on a release. Maybe I can have the binary built with xgo in parallel to tests running and only publish it in the case all tests pass... I'll see what I can do.

**Comment by nyarly** [comment]
@samsalisbury I think all the things now go into the jobs: key, so that the stages can be sequenced reliably. No docs at all about what the order of non-staged work is...

**Comment by samsalisbury** [comment]
@nyarly many of the documentation examples for jobs/stages use the top-level config followed by the `jobs:` spec, and run the top-level config first. The top-level stage gets the automatic name "Test", and we can deduce from this documentation that that `Test` stage is run first by default: https://docs.travis-ci.com/user/build-stages/#specifying-stage-order-and-conditions:

> You can specify the order for stages in the section stages: [code example removed for brevity]
>
> This is mostly useful in order to “prepend” a stage to the test stage that jobs resulting from the matrix expansion will be assigned to.

Since Travis documentation is mostly example based and light on explicit specifications, taking the above quote along with empirical evidence seems like a reasonable approach.

We can see empirically that this updated `.travis.yml` works as expected consistently on this branch, and on all the other open PRs that include these changes. See https://docs.travis-ci.com/user/build-stages/matrix-expansion/ for further example of documentation that shows the default top-level 'test' stage being run first. I don't see explicit mention that this is the case (travis docs are low on precise specs), but since it follows many of the examples in Travis, it seems like a reasonable change, even if it could potentially be improved upon in a future pull request. (We have already discussed improvements like grouping fast tests together in their own stage to fail the build quicker if those tests fail. During that change would be a good time to make the .travis.yml stages config more explicit.)

Since this has been sitting around for a while, and other work touching .travis.yml now sits on top of it, updating this branch prior to merging with master means resolving merge conflicts for little gain (most likely zero behavioural change). Therefore I would rather open a follow-up PR to make stage order more explicit rather than have to deal with these conflicts.

**Comment by samsalisbury** [comment]
Further delving into the docs suggests the test matrix has to be defined at top-level, so the best we might ultimately be able to achieve is explicit ordering of implicitly named "test" stage; see https://docs.travis-ci.com/user/build-stages/#build-stages-and-build-matrix-expansion

**Pull request merged**